pkg/memtier/tracker_finder.go

```diff
@@ -0,0 +1,269 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The finder tracker keeps reporting all memory address ranges but
+// does not track accesses. This is useful, for intance, when all
+// detected memory should be handled similarly by the policy.
+
+package memtier
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// TrackerFinderConfig represents the configuration for TrackerFinder.
+type TrackerFinderConfig struct {
+	// RegionsUpdateMs defines process memory region update
+	// interval in milliseconds. Regions are updated just before
+	// scanning pages if the interval has passed. Value 0 means
+	// that regions are updated before every scan.
+	RegionsUpdateMs uint64
+	// ReportAccesses is the number of accesses the tracker
+	// reports on every address range it has found.
+	ReportAccesses uint64
+	// PagemapBitBIT filters in pages with only given values
+	// of pagemap bits. If nil, there is no filtering.
+	// If all PagemapBitBITs are nil, finding memory regions
+	// is much faster because any pagemap bits are not checked.
+	// For more information on pagemap bits, refer to:
+	// https://www.kernel.org/doc/Documentation/vm/pagemap.txt
+	PagemapBitExclusive *bool
+	PagemapBitPresent   *bool
+	PagemapBitSoftDirty *bool
+}
+
+const trackerFinderDefaults string = `{"RegionsUpdateMs":5000}`
+
+// TrackerFinder is a memory tracker that reports address ranges
+// without any accesses on them.
+type TrackerFinder struct {
+	mutex        sync.Mutex
+	regionsMutex sync.Mutex
+	config       *TrackerFinderConfig
+	regions      map[int]*AddrRanges
+	toSampler    chan byte
+}
+
+func init() {
+	TrackerRegister("finder", NewTrackerFinder)
+}
+
+// NewTrackerFinder creates a new instance of TrackerFinder.
+func NewTrackerFinder() (Tracker, error) {
+	t := &TrackerFinder{
+		regions: make(map[int]*AddrRanges),
+	}
+	err := t.SetConfigJSON(trackerFinderDefaults)
+	if err != nil {
+		return nil, fmt.Errorf("invalid finder default configuration")
+	}
+	return t, nil
+}
+
+// SetConfigJSON sets the configuration for TrackerFinder from a JSON string.
+func (t *TrackerFinder) SetConfigJSON(configJSON string) error {
+	config := &TrackerFinderConfig{}
+	if err := unmarshal(configJSON, config); err != nil {
+		return err
+	}
+	t.config = config
+	return nil
+}
+
+// GetConfigJSON returns the JSON representation of the TrackerFinder's configuration.
+func (t *TrackerFinder) GetConfigJSON() string {
+	if t.config == nil {
+		return ""
+	}
+	if configStr, err := json.Marshal(t.config); err == nil {
+		return string(configStr)
+	}
+	return ""
+}
+
+func (t *TrackerFinder) addRanges(pid int) {
+	delete(t.regions, pid)
```

Comment:
Add t.regionsMutex.Lock() and unlock() for delete(t.regions, pid)?

Reply:
This would cause a deadlock. addRanges() is called always by contexts that have already taken t.regionsMutex.Lock(). And we probably do not want to unlock t.regionsMutex when some of the regions are updated and some not...

Reply:
ahah okay